### PR TITLE
[Feature]add support for be graceful exit

### PR DESCRIPTION
**Pull request by cbcbq**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

change graceful exit behaviour to:
1. BE tell FE it will exit soon
2. FE won't dispatch new fragments to BE exiting
3. BE loop wait all fragments done (maximum loop count is configurable), then stop servers and exit

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 28 (53.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/ComputeNode.java|7|18|38.89%|[217, 218, 295, 296, 302, 341, 349, 350, 421, 434, 435]|
|:large_blue_circle:|com/starrocks/system/HeartbeatMgr.java|6|8|75.00%|[304, 311]|
|:large_blue_circle:|com/starrocks/system/BackendHbResponse.java|2|2|100.00%|[]|

File: gensrc/thrift/HeartbeatService.thrift [hunk 1]
```diff
@@ -29,6 +29,7 @@ struct TMasterInfo {
     6: optional Types.TPort http_port
     7: optional i64 heartbeat_flags
     8: optional i64 backend_id
+    9: optional i32 backend_state
```
Comment (imay) [inline gensrc/thrift/HeartbeatService.thrift:32]: Why does leader tell backend's state?

File: gensrc/thrift/HeartbeatService.thrift [hunk 2]
```diff
@@ -39,6 +40,7 @@ struct TBackendInfo {
     5: optional string version
     6: optional i32 num_hardware_cores
     7: optional Types.TPort starlet_port
+    8: optional i32 state
```
Comment (imay) [inline gensrc/thrift/HeartbeatService.thrift:43]: should be TNodeState?

File: gensrc/thrift/Types.thrift [hunk 3]
```diff
@@ -446,3 +446,10 @@ enum TCompressionType {
     BZIP2 = 10;
     LZO = 11; // Deprecated
 }
+
+enum THbBackendState {
+    INVALID_STATE = 0,
+    ALIVE = 1,
+    PREPARE_EXIT = 2,
+    OFFLINE = 3,
+}
```
Comment (imay) [inline gensrc/thrift/Types.thrift:455]: How about naming it TNodeState, I think this state may be OK for compute node, frontend node. And their states should be
NORMAL,
COOLDOWN,
DOWN

File: be/src/service/service.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#include "service/service.h"
```
Comment (imay) [inline be/src/service/service.cpp:1]: should add license header

File: gensrc/thrift/HeartbeatService.thrift (hunk 1 above)
Comment (cbcbq) [reply gensrc/thrift/HeartbeatService.thrift:32]: means backend's state in master's view

File: be/src/service/service.cpp (hunk 4 above)
Comment (cbcbq) [reply be/src/service/service.cpp:3]: license added

**Comment by cbcbq** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 28 (53.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/ComputeNode.java|7|18|38.89%|[217, 218, 295, 296, 302, 341, 349, 350, 421, 434, 435]|
|:large_blue_circle:|com/starrocks/system/HeartbeatMgr.java|6|8|75.00%|[304, 311]|
|:large_blue_circle:|com/starrocks/system/BackendHbResponse.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 28 (53.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/ComputeNode.java|7|18|38.89%|[217, 218, 295, 296, 302, 341, 349, 350, 421, 434, 435]|
|:large_blue_circle:|com/starrocks/system/HeartbeatMgr.java|6|8|75.00%|[304, 311]|
|:large_blue_circle:|com/starrocks/system/BackendHbResponse.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/system/ComputeNode.java [hunk 5]
```diff
@@ -49,6 +50,8 @@ public class ComputeNode implements IComputable, Writable {
     private volatile long lastStartTime;
     @SerializedName("isAlive")
     private AtomicBoolean isAlive;
+    @SerializedName("isPrepareExit")
+    private AtomicBoolean isPrepareExit;
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/system/ComputeNode.java:54]: why not store a State here?

File: gensrc/thrift/HeartbeatService.thrift [hunk 6]
```diff
@@ -29,6 +29,7 @@ struct TMasterInfo {
     6: optional Types.TPort http_port
     7: optional i64 heartbeat_flags
     8: optional i64 backend_id
+    9: optional Types.TNodeState backend_state
```
[the remark this answers is not in the document]
Comment (cbcbq) [reply gensrc/thrift/HeartbeatService.thrift:32]: same as an ACK
BE can begin to exit only after master had set BE's state to COOLDOWN

[the remark this answers is not in the document]
Comment (imay) [reply gensrc/thrift/HeartbeatService.thrift:32]: Is it necessary? IMO, even without Leader ACK, BE can exit too.

**Review by imay — DISMISSED** [review]
LGTM

File: be/src/common/config.h [hunk 7]
```diff
@@ -769,4 +769,10 @@ CONF_mBool(dependency_librdkafka_debug_enable, "false");
 // admin, eos, mock, assigner, conf
 CONF_String(dependency_librdkafka_debug, "all");
 
+// max loop count when be waiting its fragments finish
+CONF_Int64(be_loop_count_wait_fragments_finish, "0");
```
Comment (kangkaisen) [inline be/src/common/config.h:773]: I think the BE and CN could use one config.

Comment (cbcbq) [reply be/src/common/config.h]: changed to loop_count_wait_fragments_finish

**Review by imay — DISMISSED** [review]

File: be/src/runtime/fragment_mgr.h [hunk 8]
```diff
@@ -88,6 +88,7 @@ class FragmentMgr : public RestMonitorIface {
     // execute external query, all query info are packed in TScanOpenParams
     Status exec_external_plan_fragment(const TScanOpenParams& params, const TUniqueId& fragment_instance_id,
                                        std::vector<TScanColumnDesc>* selected_columns);
+    size_t running_fragment_count() const { return _fragment_map.size(); }
```
Comment (kangkaisen) [inline be/src/runtime/fragment_mgr.h:91]: need the lock

**Review by kangkaisen — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**